### PR TITLE
11.0 mig stock change qty reason

### PR DESCRIPTION
**Pull request by andreasdsp**
[MIG] stock_change_qty_reason: Migration to 11.0

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please squash together adjacent "OCA Transbot.." commits using https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

**Review by gurneyalex — APPROVED** [review]
Code review

File: stock_change_qty_reason/README.rst
Comment (pedrobaeza) [inline stock_change_qty_reason/README.rst]: 11.0

File: stock_change_qty_reason/README.rst
Comment (pedrobaeza) [inline stock_change_qty_reason/README.rst]: Use latest template: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
You can also squash together the first 4 commits from the first author, as they are all a logical group.

**Comment by rousseldenis** [comment]
@andreasdsp

File: stock_change_qty_reason/wizard/stock_change_product_qty.py
Comment (rousseldenis) [inline stock_change_qty_reason/wizard/stock_change_product_qty.py:3]: Why this ?

**Review by rousseldenis — APPROVED** [review]
Minor change but LGTM

**Comment by rousseldenis** [comment]
@pedrobaeza Can you update your review?

**Comment by rousseldenis** [comment]
@pedrobaeza IMHO the transbot commits don't block this (only 2).

**Comment by pedrobaeza** [comment]
I have rebased and added README by fragments. Merging when CIs finish.

**Review by pedrobaeza — APPROVED** [review]

**Comment by JordiBForgeFlow** [comment]
@andreasdsp I do not really agree on this module. In my opinion the reasons should not be free-form, but tabulated. It should be possible to indicate the reason for the change in quantity in both the product wizard, and also when you do an inventory adjustment.

If you use stock accounting, you should be able to associate an account.account to a reason code, so that posting an inventory adjustment using a reason would post to that account instead of the default account used by the virtual location of inventory loss.

We will propose a separate module that introduces an inventory.change.reason as a separate model.

**Comment by pedrobaeza** [comment]
@jbeficent this is only a migration...

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza OK then. We will propose soon 2 new modules:
stock_inventory_change_reason
stock_account_inventory_change_reason

**Comment by pedrobaeza** [comment]
But if they are overlapping features, then it shouldn't be both. Why don't you expand this one allowing to have predefined texts? You have 2 fields: the many2one and the text, filling the text one with the name of the many2one record and making the many2one not required.

About `stock_account_inventory_change_reason`, I don't get the goal.

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza The goal is that, for example, an inventory adjustment made with reason 'Expenses by marketing' should post the inventory loss to the account 'Marketing expenses', instead of expensing to the generic 'Inventory loss' account.

If you do not manage perpetual inventory this won't apply, of course.

**Comment by JordiBForgeFlow** [comment]
> But if they are overlapping features, then it shouldn't be both. Why don't you expand this one allowing to have predefined texts? You have 2 fields: the many2one and the text, filling the text one with the name of the many2one record and making the many2one not required.
> 
> About `stock_account_inventory_change_reason`, I don't get the goal.

Putting an encoded reason is important, because it allows the organization to control why people make changes to inventory. It is a best practice to have a catalog of reason codes:

Microsoft: https://docs.microsoft.com/en-us/dynamics365/unified-operations/supply-chain/warehousing/reason-codes-for-counting-journals

SAP: https://help.sap.com/SAPhelp_47x200/helpdata/en/68/f51f8a12692a4d9a5c0304e4cca87a/content.htm?no_cache=true

**Comment by JordiBForgeFlow** [comment]
Entering a reason as a many2one, with reasons being a separate model, allow also to add additional behaviour, such as what account.account to use as contra-account for the inventory adjustment. This is also a best practive in inventory management

**Comment by pedrobaeza** [comment]
@jbeficent I'm not against the feature, but I'm talking about integrating the feature in current module the way I told.

**Comment by JordiBForgeFlow** [comment]
The problem is that, as it is now, it is based on a Char, which makes it hard to extend. But we will analyze if we can find a way without making the module too complex.

**Comment by pedrobaeza** [comment]
Please re-read my explanations about how to make it work with both fields.

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza Ok, I follow you now.

**Review by HviorForgeFlow — APPROVED** [review]

File: stock_change_qty_reason/wizard/stock_change_product_qty.py
Comment (JordiBForgeFlow) [inline stock_change_qty_reason/wizard/stock_change_product_qty.py:21]: sending the info using context is not a good idea, as I discussed. It is better to extend the stock.inventory.line with a field 'reason' and pass the value from the wizard using https://github.com/odoo/odoo/blob/11.0/addons/stock/wizard/stock_change_product_qty.py#L50

File: stock_change_qty_reason/models/stock_inventory_line.py
Comment (JordiBForgeFlow) [inline stock_change_qty_reason/models/stock_inventory_line.py:11]: define a field 'reason' in stock.inventory.line, and fetch the value, to then send it to the stock move.

**Review by JordiBForgeFlow — CHANGES_REQUESTED** [review]
We need an approach to indicate the justification for changes that can be updated both in the wizard 'stock.change.product.qty' and in the stock inventory adjustment. We are working on that already in https://github.com/OCA/stock-logistics-warehouse/pull/524

File: stock_change_qty_reason/wizard/stock_change_product_qty.py
Comment (rousseldenis) [reply stock_change_qty_reason/wizard/stock_change_product_qty.py:21]: I agree despite if I did that in the original module. Bad implementation. :smiley:

**Comment by JordiBForgeFlow** [comment]
@rousseldenis @gurneyalex @andreasdsp @pedrobaeza we have created the proposed PR here: #524 . I think that it's better that we review the newer PR before we merge. But happy also to merge now this one and continue with #524.

**Comment by rousseldenis** [comment]
@jbeficent Ok, let's merge this and continue with #524 

Do I consider your request changes to be applied on new PR ?

**Comment by LoisRForgeFlow** [comment]
@rousseldenis yes, you can review the other PR, it is also ready.

**Pull request merged**
